File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/detail/KanjiDetailViewModel.kt
package com.jworks.kanjiquest.android.ui.detail

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.android.data.PreviewTrialManager
import com.jworks.kanjiquest.core.domain.UserSessionProvider
import com.jworks.kanjiquest.core.domain.model.GameMode
import com.jworks.kanjiquest.core.domain.model.Kanji
import com.jworks.kanjiquest.core.domain.model.UserLevel
import com.jworks.kanjiquest.core.domain.model.Vocabulary
import com.jworks.kanjiquest.core.domain.repository.FlashcardRepository
import com.jworks.kanjiquest.core.domain.repository.KanjiRepository
import com.jworks.kanjiquest.core.domain.repository.SrsRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ModeTrialInfo(
    val canPractice: Boolean,
    val trialsRemaining: Int
)

data class KanjiDetailUiState(
    val kanji: Kanji? = null,
    val vocabulary: List<Vocabulary> = emptyList(),
    val isLoading: Boolean = true,
    val totalPracticeCount: Int = 0,
    val accuracy: Float? = null,
    val isInFlashcardDeck: Boolean = false,
    val canPracticeWriting: Boolean = false,
    val writingTrialsRemaining: Int = 0,
    val isPremium: Boolean = false,
    val isAdmin: Boolean = false,
    val modeTrials: Map<GameMode, ModeTrialInfo> = emptyMap(),
    val modeStats: Map<String, Int> = emptyMap()
)

@HiltViewModel
class KanjiDetailViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val kanjiRepository: KanjiRepository,
    private val srsRepository: SrsRepository,
    private val flashcardRepository: FlashcardRepository,
    private val userSessionProvider: UserSessionProvider,
    private val previewTrialManager: PreviewTrialManager
) : ViewModel() {

    val kanjiId: Int = checkNotNull(savedStateHandle["kanjiId"])

    private val _uiState = MutableStateFlow(KanjiDetailUiState())
    val uiState: StateFlow<KanjiDetailUiState> = _uiState.asStateFlow()

    init {
        loadKanji()
    }

    private fun loadKanji() {
        viewModelScope.launch {
            val kanji = kanjiRepository.getKanjiById(kanjiId)
            val vocab = if (kanji != null) {
                kanjiRepository.getVocabularyForKanji(kanjiId)
            } else emptyList()

            val srsCard = srsRepository.getCard(kanjiId)
            val totalPractice = srsCard?.totalReviews ?: 0
            val accuracy = if (srsCard != null && srsCard.totalReviews > 0) srsCard.accuracy else null

            val inDeck = flashcardRepository.isInDeck(kanjiId)

            val effectiveLevel = userSessionProvider.getEffectiveLevel()
            val isPremium = effectiveLevel == UserLevel.PREMIUM || effectiveLevel == UserLevel.ADMIN
            val isAdmin = userSessionProvider.isAdmin()
            val writingTrials = previewTrialManager.getRemainingTrials(GameMode.WRITING)
            val canPractice = isPremium || isAdmin || writingTrials > 0

            // Per-mode trial info
            val modeTrials = mapOf(
                GameMode.RECOGNITION to ModeTrialInfo(canPractice = true, trialsRemaining = -1),
                GameMode.WRITING to ModeTrialInfo(
                    canPractice = isPremium || isAdmin || writingTrials > 0,
                    trialsRemaining = writingTrials
                ),
                GameMode.VOCABULARY to run {
                    val trials = previewTrialManager.getRemainingTrials(GameMode.VOCABULARY)
                    ModeTrialInfo(canPractice = isPremium || isAdmin || trials > 0, trialsRemaining = trials)
                },
                GameMode.CAMERA_CHALLENGE to run {
                    val trials = previewTrialManager.getRemainingTrials(GameMode.CAMERA_CHALLENGE)
                    ModeTrialInfo(canPractice = isPremium || isAdmin || trials > 0, trialsRemaining = trials)
                }
            )

            // Per-mode stats
            val modeStats = try {
                val statsMap = srsRepository.getModeStatsByIds(listOf(kanjiId.toLong()))
                statsMap[kanjiId] ?: emptyMap()
            } catch (_: Exception) { emptyMap() }

            _uiState.value = KanjiDetailUiState(
                kanji = kanji,
                vocabulary = vocab,
                isLoading = false,
                totalPracticeCount = totalPractice,
                accuracy = accuracy,
                isInFlashcardDeck = inDeck,
                canPracticeWriting = canPractice,
                writingTrialsRemaining = writingTrials,
                isPremium = isPremium,
                isAdmin = isAdmin,
                modeTrials = modeTrials,
                modeStats = modeStats
            )
        }
    }

    fun toggleFlashcard() {
        viewModelScope.launch {
            val nowInDeck = flashcardRepository.toggleInDeck(kanjiId)
            _uiState.value = _uiState.value.copy(isInFlashcardDeck = nowInDeck)
        }
    }

    fun useWritingTrial(): Boolean {
        return useModeTrial(GameMode.WRITING)
    }

    fun useModeTrial(mode: GameMode): Boolean {
        if (mode == GameMode.RECOGNITION) return true // Always free
        val success = previewTrialManager.usePreviewTrial(mode)
        if (success) {
            val remaining = previewTrialManager.getRemainingTrials(mode)
            val isPremium = _uiState.value.isPremium
            val isAdmin = _uiState.value.isAdmin
            val updatedTrials = _uiState.value.modeTrials.toMutableMap()
            updatedTrials[mode] = ModeTrialInfo(
                canPractice = isPremium || isAdmin || remaining > 0,
                trialsRemaining = remaining
            )
            _uiState.value = _uiState.value.copy(
                modeTrials = updatedTrials,
                writingTrialsRemaining = if (mode == GameMode.WRITING) remaining else _uiState.value.writingTrialsRemaining,
                canPracticeWriting = if (mode == GameMode.WRITING) (isPremium || isAdmin || remaining > 0) else _uiState.value.canPracticeWriting
            )
        }
        return success
    }
}

File: ios-app/KanjiQuest/Screens/Detail/KanjiDetailViewModel.swift
import Foundation
import SharedCore

struct ModeTrialInfo {
    let canPractice: Bool
    let trialsRemaining: Int
}

@MainActor
final class KanjiDetailViewModel: ObservableObject {
    @Published var kanji: Kanji? = nil
    @Published var vocabulary: [Vocabulary] = []
    @Published var vocabSentences: [Int64: ExampleSentence] = [:]
    @Published var isLoading = true
    @Published var totalPracticeCount = 0
    @Published var accuracy: Float? = nil
    @Published var isInFlashcardDeck = false
    @Published var isPremium = false
    @Published var isAdmin = false
    @Published var modeTrials: [String: ModeTrialInfo] = [:]
    @Published var showDeckChooser = false
    @Published var deckGroups: [FlashcardDeckGroup] = []
    @Published var kanjiInDecks: [Int64] = []

    private var kanjiRepository: KanjiRepository?
    private var srsRepository: SrsRepository?
    private var flashcardRepository: FlashcardRepository?
    private var userRepository: UserRepository?
    private var previewTrialManager: PreviewTrialManager?

    func load(container: AppContainer, kanjiId: Int32) {
        kanjiRepository = container.kanjiRepository
        srsRepository = container.srsRepository
        flashcardRepository = container.flashcardRepository
        userRepository = container.userRepository
        previewTrialManager = container.previewTrialManager

        Task {
            let k = try? await kanjiRepository?.getKanjiById(id: kanjiId)
            let vocab: [Vocabulary] = (k != nil) ? ((try? await kanjiRepository?.getVocabularyForKanji(kanjiId: kanjiId)) ?? []) : []

            // Load example sentences
            var sentences: [Int64: ExampleSentence] = [:]
            for v in vocab.prefix(10) {
                if let sentence = try? await kanjiRepository?.getExampleSentence(vocabId: v.id) {
                    sentences[v.id] = sentence
                }
            }

            // SRS stats
            let srsCard = try? await srsRepository?.getCard(kanjiId: kanjiId)
            let totalPractice = Int(srsCard?.totalReviews ?? 0)
            let acc: Float? = (srsCard != nil && srsCard!.totalReviews > 0) ? srsCard!.accuracy : nil

            // Flashcard state
            let inDeckRaw = try? await flashcardRepository?.isInDeck(kanjiId: kanjiId)
            let inDeck = (inDeckRaw as? NSNumber)?.boolValue ?? false

            // Premium/admin check
            let premium = container.userSessionProvider.isPremium()
            let admin = container.userSessionProvider.isAdmin()

            // Mode trials
            let writingTrials = previewTrialManager?.getRemainingTrials(mode: "WRITING") ?? 0
            let vocabTrials = previewTrialManager?.getRemainingTrials(mode: "VOCABULARY") ?? 0
            let cameraTrials = previewTrialManager?.getRemainingTrials(mode: "CAMERA_CHALLENGE") ?? 0

            let trials: [String: ModeTrialInfo] = [
                "recognition": ModeTrialInfo(canPractice: true, trialsRemaining: -1),
                "writing": ModeTrialInfo(canPractice: premium || admin || writingTrials > 0, trialsRemaining: writingTrials),
                "vocabulary": ModeTrialInfo(canPractice: premium || admin || vocabTrials > 0, trialsRemaining: vocabTrials),
                "camera_challenge": ModeTrialInfo(canPractice: premium || admin || cameraTrials > 0, trialsRemaining: cameraTrials)
            ]

            self.kanji = k
            self.vocabulary = vocab
            self.vocabSentences = sentences
            self.totalPracticeCount = totalPractice
            self.accuracy = acc
            self.isInFlashcardDeck = inDeck
            self.isPremium = premium
            self.isAdmin = admin
            self.modeTrials = trials
            self.isLoading = false
        }
    }

    func toggleFlashcard() {
        Task {
            try? await flashcardRepository?.ensureDefaultDeck()
            let groups = (try? await flashcardRepository?.getAllDeckGroups()) ?? []
            if groups.count <= 1 {
                let toggleRaw = try? await flashcardRepository?.toggleInDeck(kanjiId: kanji?.id ?? 0)
                let nowInDeck = (toggleRaw as? NSNumber)?.boolValue ?? false
                isInFlashcardDeck = nowInDeck
            } else {
                let decksForKanji = (try? await flashcardRepository?.getDecksForKanji(kanjiId: kanji?.id ?? 0)) ?? []
                deckGroups = groups
                kanjiInDecks = decksForKanji.map { Int64(truncating: $0) }
                showDeckChooser = true
            }
        }
    }

    func addToDeck(_ deckId: Int64) {
        guard let kanjiId = kanji?.id else { return }
        Task {
            try? await flashcardRepository?.addToDeck(deckId: deckId, kanjiId: kanjiId)
            let decks = (try? await flashcardRepository?.getDecksForKanji(kanjiId: kanjiId)) ?? []
            kanjiInDecks = decks.map { Int64(truncating: $0) }
            isInFlashcardDeck = true
        }
    }

    func removeFromDeck(_ deckId: Int64) {
        guard let kanjiId = kanji?.id else { return }
        Task {
            try? await flashcardRepository?.removeFromDeck(deckId: deckId, kanjiId: kanjiId)
            let decks = (try? await flashcardRepository?.getDecksForKanji(kanjiId: kanjiId)) ?? []
            kanjiInDecks = decks.map { Int64(truncating: $0) }
            isInFlashcardDeck = !decks.isEmpty
        }
    }

    func dismissDeckChooser() {
        showDeckChooser = false
    }

    func useModeTrial(mode: String) -> Bool {
        if mode == "recognition" { return true }
        guard let manager = previewTrialManager else { return false }
        let success = manager.usePreviewTrial(mode: mode)
        if success {
            let remaining = manager.getRemainingTrials(mode: mode)
            var updatedTrials = modeTrials
            updatedTrials[mode] = ModeTrialInfo(
                canPractice: isPremium || isAdmin || remaining > 0,
                trialsRemaining: remaining
            )
            modeTrials = updatedTrials
        }
        return success
    }
}
